File: numerical_methods/roots/bisection.f90
! Programa elaborado por Jesús Eduardo Loera Casas
! Fecha 04/02/23
 
        real function f(x)
            implicit none 
            real :: x
            f = sqrt(x+1)*( (cos(x/2.0))**3 )
            return
        end  

        subroutine bisection(a, b, root, step)
            implicit none
            real :: a, b, c, root, dx, f, tol
            integer :: steps, step, i
            ! definimos una tolerancia para la convergencia
            tol = 0.0001 
            ! definimos un número iteraciones maxima para la convergencia
            steps = 1000
            do i = 1, steps
                c = 0.5*(a+b)
                dx = b-a 

                if ( dx .le. tol ) then
                    step = i
                    root = c 
                    exit
                end if

                if ( f(a)*f(c)>0 ) then
                    a = c
                else if ( f(a)*f(c)<0 ) then
                    b = c
                else 
                    if ( f(a)==0 ) then
                        write(*,*) "Hay una raiz en x = ", a
                        step = i
                        root = a
                        exit
                    end if
                    if ( f(c)==0 ) then
                        write(*,*) "Hay una raiz en x = ", c 
                        step = i
                        root = c 
                        exit
                    end if
                end if
            end do
            if ( i == steps ) then
                write(*,*) "No se encontro raiz que satisfaga la"
                write(*,*) "tolerancia", tol, " en ", steps, " pasos" 
            end if
    
        end subroutine bisection


        program main
 
            implicit none
            real :: a, b, root
            integer :: step

            ! intervalo de inspección
            a = 0.2 ; b = 4.0

            call bisection(a, b, root, step)

            write(*,*) "Se encontro una raiz en ", step," iteraciones"
            write(*,*) "La raiz enocntrada es x = ", root

        end program main

File: numerical_methods/roots/bisection/bisection.py
# Programa elaborado por Jesús Eduardo Loera Casas
# Elaborado el día dd/mm/aa
 
# importamos librerías 
import numpy as np 

"""
En este programa implementamos el método de bisección para determinar 
las raíces de una función continua f(x) en algún intervalo [a,b]
"""

"""
Función continua
"""
def f(x):
    return np.sin(x)

"""
Intervalo de inspección
"""
a = 0.1
b = 2*np.pi

"""
Método de bisección
"""
def bisection(f, a, b):
    # si la raíz está entre a y b, esperamos que b-a < tolerancia
    tolerancia = 0.0001            
    dx = b - a
    # contaremos los pasos en los que se ecnuetra la primera raíz
    step = 0
    while(0 < 1):
        roots = []
        step += 1
        c = 0.5*(a+b)
        dx = b - a
        if(dx <= tolerancia):
            roots.append(c)
            break
        if (f(a)*f(c) > 0):
            a = c
        elif(f(a)*f(c) < 0):
            b = c
        else:
            if(f(c)==0):
                roots.append(c)
            if(f(a)==0):
                roots.append(a)
            break
        if (step > 10000):
            break
    return roots, step

roots = bisection(f, a, b)
print(roots)



